### PR TITLE
Make enumeration creation methods thread-safe

### LINKED ISSUE #2526: NSSet ObjectsWithOptionsPassingTest::Test Fails ~0.1% of times run
Repro steps:
Run tests as `.\TE.exe .\Foundation.UnitTests.dll /name:NSSet_ObjectsWithOptionsP* /testmode:Loop /LoopTest:10000`
Output:
Non-passing Tests:

    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]
    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]
    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]
    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]
    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]
    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]
    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]
    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]
    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]
    NSSet_ObjectsWithOptionsPassingTest::Test [Failed]

Summary: Total=10000, Passed=9990, Failed=10, Blocked=0, Not Run=0, Skipped=0

### PR DESCRIPTION
**Pull request by aballway**
Fixes #2526

### CODE REVIEW

File: tests/unittests/Foundation/NSSetTests.mm [hunk 1]
```diff
@@ -213,18 +214,27 @@
                                       --waitingCount;
                                       [condition signal];
                                       [condition unlock];
-                                      return [obj unsignedIntegerValue] % 2 == 0 ? YES : NO;
+                                      if ([obj unsignedIntegerValue] % 2 == 0) {
+                                          ++timesHit;
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSSetTests.mm]: is this threadsafe? :P

Comment (aballway) [inline tests/unittests/Foundation/NSSetTests.mm]: Whoops didn't mean to leave this, was just in there for my testing.

File: tests/unittests/Foundation/NSSetTests.mm [hunk 2]
```diff
@@ -214,16 +213,10 @@
                                       --waitingCount;
                                       [condition signal];
                                       [condition unlock];
-                                      if ([obj unsignedIntegerValue] % 2 == 0) {
-                                          ++timesHit;
-                                          return YES;
-                                      }
-
-                                      return NO;
+                                      return [obj unsignedIntegerValue] % 2 == 0 ? YES : NO;
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSSetTests.mm]: Why the `? YES : NO`? It seems superfluous.

Comment (aballway) [inline tests/unittests/Foundation/NSSetTests.mm]: The compiler was complaining about not being able to convert bool^(id,BOOL*) to BOOL^(id,BOOL*) so I don't think it's superfluous.

**Comment by bviglietta** [comment]
>     _enumerateWithBlock(enumerator, options, ^(id key, BOOL* stop) {

Looks like there are a bunch of other places where concurrent enumeration can cause problems.

---
Refers to: Frameworks/Foundation/NSArray.mm:767 in 5e4f98c. [](commit_id = 5e4f98c05bb3c2093785a53a3e2a62351fff75dc, deletion_comment = False)

**Comment by bviglietta** [comment]
>     _enumerateWithBlock(enumerator, options, ^(id key, BOOL* stop) {

NSIndexSet -indexesInRange:passingTest: would be another.

---
In reply to: [294629355](https://github.com/Microsoft/WinObjC/pull/2542#issuecomment-294629355) [](ancestors = 294629355)

---
Refers to: Frameworks/Foundation/NSArray.mm:767 in 5e4f98c. [](commit_id = 5e4f98c05bb3c2093785a53a3e2a62351fff75dc, deletion_comment = False)

**Comment by bviglietta** [comment]
:shipit:

**Pull request merged**